File: backend/kernel/scheduler.py
"""多 Agent 调度器雏形。

.. deprecated:: P0-A / R3 去双轨
    **生产权威**：Rust ``scheduler_*`` / ``run_gate_*`` RPC。
    本文件仅 fallback / 历史测试；禁止扩展生产队列逻辑。

H-07：Session 锁只保证同会话不重入，不是全局调度器；
跨会话排队与前台优先由 run_gate + priority_class 完成。
"""

from __future__ import annotations

import heapq
import itertools
import time
import uuid
from dataclasses import dataclass, field
from typing import Any

_AGE_THRESHOLD_SECONDS = 30.0  # 等待超过 30s 提一档
_AGE_BOOST = 1  # 每次提升的优先级档数


@dataclass(order=True)
class ScheduledTask:
    """优先级队列元素。heapq 按 (effective_priority, seq) 排序。"""

    effective_priority: int
    seq: int  # 同优先级 FIFO 序号
    id: str = field(compare=False, default_factory=lambda: uuid.uuid4().hex[:12])
    process_id: str = field(compare=False, default="")
    payload: dict[str, Any] = field(compare=False, default_factory=dict)
    base_priority: int = field(compare=False, default=10)
    submitted_at: float = field(compare=False, default_factory=time.time)
    state: str = field(compare=False, default="queued")  # queued/running/done/cancelled


class AgentScheduler:
    """优先级调度器（雏形）。

    线程安全由调用方保证（asyncio 单线程语义下无需锁）。
    """

    def __init__(self, *, age_threshold: float = _AGE_THRESHOLD_SECONDS) -> None:
        self._heap: list[ScheduledTask] = []
        self._tasks: dict[str, ScheduledTask] = {}
        self._seq = itertools.count()
        self._age_threshold = age_threshold

    def submit(
        self,
        process_id: str,
        payload: dict[str, Any] | None = None,
        *,
        priority: int = 10,
    ) -> ScheduledTask:
        task = ScheduledTask(
            effective_priority=max(0, priority),
            seq=next(self._seq),
            process_id=process_id,
            payload=dict(payload or {}),
            base_priority=max(0, priority),
        )
        self._tasks[task.id] = task
        heapq.heappush(self._heap, task)
        return task

    def _apply_aging(self) -> None:
        """等待过久的任务提升优先级（重建堆）。O(n)，任务量小可接受。"""
        now = time.time()
        dirty = False
        for t in self._tasks.values():
            if t.state != "queued":
                continue
            waited = now - t.submitted_at
            boost = int(waited // self._age_threshold) * _AGE_BOOST
            new_prio = max(0, t.base_priority - boost)
            if new_prio != t.effective_priority:
                t.effective_priority = new_prio
                dirty = True
        if dirty:
            heapq.heapify(self._heap)

    def next(self) -> ScheduledTask | None:
        """取出下一个应执行的任务（标记 running）。"""
        self._apply_aging()
        while self._heap:
            task = heapq.heappop(self._heap)
            if task.id in self._tasks and task.state == "queued":
                task.state = "running"
                return task
        return None

    def complete(self, task_id: str, *, cancelled: bool = False) -> None:
        task = self._tasks.get(task_id)
        if task is not None:
            task.state = "cancelled" if cancelled else "done"
            # 终态记录有界保留（供 stats/审计），超上限清最旧
            terminal = [t for t in self._tasks.values() if t.state in ("done", "cancelled")]
            if len(terminal) > 1000:
                for t in sorted(terminal, key=lambda x: x.seq)[: len(terminal) - 1000]:
                    self._tasks.pop(t.id, None)

    def cancel_process(self, process_id: str) -> int:
        """进程终止时取消其所有排队任务。返回取消数。"""
        n = 0
        for t in self._tasks.values():
            if t.process_id == process_id and t.state == "queued":
                t.state = "cancelled"
                n += 1
        return n

    def queued(self) -> list[ScheduledTask]:
        self._apply_aging()
        return sorted(
            (t for t in self._tasks.values() if t.state == "queued"),
            key=lambda t: (t.effective_priority, t.seq),
        )

    def stats(self) -> dict[str, int]:
        out: dict[str, int] = {"queued": 0, "running": 0, "done": 0, "cancelled": 0}
        for t in self._tasks.values():
            out[t.state] = out.get(t.state, 0) + 1
        return out

File: crates/tevarn-kernel/src/scheduler.rs
//! Multi-agent priority scheduler with aging fairness.

use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

const AGE_THRESHOLD_SECONDS: f64 = 30.0;
const AGE_BOOST: i32 = 1;

/// Named priority classes (lower number = higher priority for heap).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PriorityClass {
    System = 0,
    Foreground = 5,
    Interactive = 10,
    WorkforceHigh = 20,
    Workforce = 30,
    Background = 50,
}

impl PriorityClass {
    pub fn as_i32(self) -> i32 {
        self as i32
    }

    pub fn parse(s: &str) -> Self {
        match s.to_ascii_lowercase().as_str() {
            "system" => Self::System,
            "foreground" | "owner" | "chat" => Self::Foreground,
            "interactive" => Self::Interactive,
            "workforce_high" | "high" => Self::WorkforceHigh,
            "workforce" | "normal" => Self::Workforce,
            "background" | "cron" | "low" => Self::Background,
            _ => Self::Workforce,
        }
    }
}

fn now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

fn short_id() -> String {
    uuid::Uuid::new_v4().simple().to_string()[..12].to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScheduledTask {
    pub id: String,
    pub process_id: String,
    pub payload: Value,
    pub base_priority: i32,
    pub effective_priority: i32,
    pub seq: u64,
    pub submitted_at: f64,
    pub state: String, // queued | running | done | cancelled
}

#[derive(Eq, PartialEq)]
struct HeapEntry {
    effective_priority: i32,
    seq: u64,
    id: String,
}

impl Ord for HeapEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        // min-heap by priority then seq
        other
            .effective_priority
            .cmp(&self.effective_priority)
            .then_with(|| other.seq.cmp(&self.seq))
    }
}

impl PartialOrd for HeapEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[derive(Default)]
pub struct AgentScheduler {
    heap: BinaryHeap<HeapEntry>,
    tasks: HashMap<String, ScheduledTask>,
    seq: u64,
    age_threshold: f64,
    /// global max concurrent running tasks (OS-level gate, independent of session locks)
    max_running: usize,
    running_count: usize,
    /// session_id → running count (fair share, not a hard session mutex)
    running_by_session: HashMap<String, usize>,
    max_per_session: usize,
    completed: u64,
    starved_skips: u64,
}

impl AgentScheduler {
    pub fn new() -> Self {
        Self {
            age_threshold: AGE_THRESHOLD_SECONDS,
            max_running: 16,
            max_per_session: 2,
            ..Default::default()
        }
    }

    pub fn with_age_threshold(mut self, t: f64) -> Self {
        self.age_threshold = t;
        self
    }

    pub fn set_limits(&mut self, max_running: usize, max_per_session: usize) {
        self.max_running = max_running.max(1);
        self.max_per_session = max_per_session.max(1);
    }

    fn session_of(payload: &Value) -> String {
        payload
            .get("session_id")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string()
    }

    pub fn submit(&mut self, process_id: impl Into<String>, payload: Value, priority: i32) -> ScheduledTask {
        let priority = priority.max(0);
        let seq = self.seq;
        self.seq += 1;
        let task = ScheduledTask {
            id: short_id(),
            process_id: process_id.into(),
            payload,
            base_priority: priority,
            effective_priority: priority,
            seq,
            submitted_at: now_secs(),
            state: "queued".into(),
        };
        self.heap.push(HeapEntry {
            effective_priority: task.effective_priority,
            seq: task.seq,
            id: task.id.clone(),
        });
        self.tasks.insert(task.id.clone(), task.clone());
        task
    }

    fn apply_aging(&mut self) {
        let now = now_secs();
        let mut dirty = false;
        for t in self.tasks.values_mut() {
            if t.state != "queued" {
                continue;
            }
            let waited = now - t.submitted_at;
            let boost = ((waited / self.age_threshold).floor() as i32) * AGE_BOOST;
            let new_prio = (t.base_priority - boost).max(0);
            if new_prio != t.effective_priority {
                t.effective_priority = new_prio;
                dirty = true;
            }
        }
        if dirty {
            self.heap.clear();
            for t in self.tasks.values() {
                if t.state == "queued" {
                    self.heap.push(HeapEntry {
                        effective_priority: t.effective_priority,
                        seq: t.seq,
                        id: t.id.clone(),
                    });
                }
            }
        }
    }

    pub fn next(&mut self) -> Option<ScheduledTask> {
        self.apply_aging();
        if self.running_count >= self.max_running {
            return None;
        }
        let mut deferred: Vec<HeapEntry> = Vec::new();
        let mut result = None;
        while let Some(entry) = self.heap.pop() {
            let Some(task) = self.tasks.get(&entry.id) else {
                continue;
            };
            if task.state != "queued" {
                continue;
            }
            let sid = Self::session_of(&task.payload);
            if !sid.is_empty() {
                let srun = *self.running_by_session.get(&sid).unwrap_or(&0);
                if srun >= self.max_per_session {
                    // fair share: skip this session's extra work, try next
                    deferred.push(entry);
                    self.starved_skips = self.starved_skips.saturating_add(1);
                    continue;
                }
            }
            if let Some(task) = self.tasks.get_mut(&entry.id) {
                task.state = "running".into();
                self.running_count = self.running_count.saturating_add(1);
                if !sid.is_empty() {
                    *self.running_by_session.entry(sid).or_insert(0) += 1;
                }
                result = Some(task.clone());
            }
            break;
        }
        // put deferred back
        for e in deferred {
            self.heap.push(e);
        }
        result
    }

    pub fn complete(&mut self, task_id: &str, cancelled: bool) {
        if let Some(task) = self.tasks.get_mut(task_id) {
            if task.state == "running" {
                self.running_count = self.running_count.saturating_sub(1);
                let sid = Self::session_of(&task.payload);
                if !sid.is_empty() {
                    if let Some(n) = self.running_by_session.get_mut(&sid) {
                        *n = n.saturating_sub(1);
                    }
                }
            }
            task.state = if cancelled {
                "cancelled".into()
            } else {
                "done".into()
            };
            self.completed = self.completed.saturating_add(1);
        }
        let terminal: Vec<_> = self
            .tasks
            .values()
            .filter(|t| t.state == "done" || t.state == "cancelled")
            .map(|t| (t.seq, t.id.clone()))
            .collect();
        if terminal.len() > 1000 {
            let mut sorted = terminal;
            sorted.sort_by_key(|(seq, _)| *seq);
            let drop_n = sorted.len() - 1000;
            for (_, id) in sorted.into_iter().take(drop_n) {
                self.tasks.remove(&id);
            }
        }
    }

    pub fn cancel_process(&mut self, process_id: &str) -> usize {
        let mut n = 0;
        for t in self.tasks.values_mut() {
            if t.process_id == process_id && t.state == "queued" {
                t.state = "cancelled".into();
                n += 1;
            }
        }
        n
    }

    pub fn queued(&mut self) -> Vec<ScheduledTask> {
        self.apply_aging();
        let mut out: Vec<_> = self
            .tasks
            .values()
            .filter(|t| t.state == "queued")
            .cloned()
            .collect();
        out.sort_by(|a, b| {
            a.effective_priority
                .cmp(&b.effective_priority)
                .then(a.seq.cmp(&b.seq))
        });
        out
    }

    pub fn stats(&self) -> HashMap<String, usize> {
        let mut out = HashMap::from([
            ("queued".into(), 0usize),
            ("running".into(), 0),
            ("done".into(), 0),
            ("cancelled".into(), 0),
            ("max_running".into(), self.max_running),
            ("max_per_session".into(), self.max_per_session),
            ("running_count".into(), self.running_count),
            ("completed_total".into(), self.completed as usize),
            ("starved_skips".into(), self.starved_skips as usize),
            ("sessions_active".into(), self.running_by_session.len()),
        ]);
        for t in self.tasks.values() {
            *out.entry(t.state.clone()).or_insert(0) += 1;
        }
        out
    }

    pub fn status(&self) -> Value {
        json!({
            "stats": self.stats(),
            "fair_share": true,
            "session_lock_decoupled": true,
            "global_cap": self.max_running,
            "per_session_cap": self.max_per_session,
            "authority": "rust",
            "os_level": true,
        })
    }
}
